File: frontend/src/services/operations.ts
/**
 * Operations Intelligence API client.
 */

import { apiFetch } from "@/services/api-client";

export type OperationsMetric = {
  id: string;
  label: string;
  value: number | null;
  formatted_value: string | null;
  unit: string | null;
  delta: number | null;
  delta_label: string | null;
  trend: "up" | "down" | "flat" | null;
  format: "currency" | "percent" | "number" | null;
  available: boolean;
  source: string | null;
};

export type OperationsOverview = {
  metrics: OperationsMetric[];
  period_start: string | null;
  period_end: string | null;
  generated_at: string;
};

export type InventoryRow = {
  product: string;
  product_id: string | null;
  sku: string | null;
  category: string | null;
  stock: number | null;
  safety_stock: number | null;
  turnover: number | null;
  inventory_value: number | null;
  stock_status: string | null;
};

export type InventoryPage = {
  rows: InventoryRow[];
  pagination: {
    page: number;
    page_size: number;
    total_items: number;
    total_pages: number;
  };
  available: boolean;
};

export type SupplierPerformanceRow = {
  supplier: string;
  supplier_id: string | null;
  on_time_pct: number | null;
  quality_score: number | null;
  lead_time: number | null;
  purchase_volume: number | null;
  risk_level: string | null;
};

export type SupplierPerformance = {
  rows: SupplierPerformanceRow[];
  available: boolean;
};

export type ReturnsRow = {
  category: string;
  return_count: number | null;
  return_pct: number | null;
  return_cost: number | null;
  trend: "up" | "down" | "flat" | null;
};

export type Returns = {
  rows: ReturnsRow[];
  available: boolean;
};

export type WarehousePerformanceRow = {
  warehouse: string;
  inventory: number | null;
  fulfillment: number | null;
  stockouts: number | null;
  average_processing_time: number | null;
};

export type WarehousePerformance = {
  rows: WarehousePerformanceRow[];
  available: boolean;
};

export type OperationalRiskRow = {
  risk: string;
  severity: "low" | "medium" | "high" | "critical";
  owner: string | null;
  recommendation: string | null;
};

export type OperationalRisks = {
  rows: OperationalRiskRow[];
  available: boolean;
};

export type OperationsQueryParams = {
  dateFrom?: string;
  dateTo?: string;
  regionIds?: string[];
  categoryIds?: string[];
  supplierIds?: string[];
  search?: string;
};

function toQuery(params: Record<string, string | number | undefined | null>): string {
  const sp = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === null || value === "") continue;
    sp.set(key, String(value));
  }
  const q = sp.toString();
  return q ? `?${q}` : "";
}

function filterParams(filters: OperationsQueryParams) {
  return {
    date_from: filters.dateFrom || undefined,
    date_to: filters.dateTo || undefined,
    region: filters.regionIds?.length ? filters.regionIds.join(",") : undefined,
    category: filters.categoryIds?.length ? filters.categoryIds.join(",") : undefined,
    supplier: filters.supplierIds?.length ? filters.supplierIds.join(",") : undefined,
    search: filters.search || undefined,
  };
}

const BASE = "/api/v1/operations";

export const operationsApi = {
  overview: (filters: OperationsQueryParams = {}) =>
    apiFetch<OperationsOverview>(`${BASE}/overview${toQuery(filterParams(filters))}`),
  inventory: (
    options: OperationsQueryParams & {
      page?: number;
      pageSize?: number;
      sortBy?: string;
      sortDir?: "asc" | "desc";
    } = {},
  ) =>
    apiFetch<InventoryPage>(
      `${BASE}/inventory${toQuery({
        ...filterParams(options),
        page: options.page ?? 1,
        page_size: options.pageSize ?? 10,
        sort_by: options.sortBy ?? "inventory_value",
        sort_dir: options.sortDir ?? "desc",
      })}`,
    ),
  supplierPerformance: (filters: OperationsQueryParams = {}) =>
    apiFetch<SupplierPerformance>(
      `${BASE}/supplier-performance${toQuery(filterParams(filters))}`,
    ),
  returns: (filters: OperationsQueryParams = {}) =>
    apiFetch<Returns>(`${BASE}/returns${toQuery(filterParams(filters))}`),
  warehousePerformance: (filters: OperationsQueryParams = {}) =>
    apiFetch<WarehousePerformance>(
      `${BASE}/warehouse-performance${toQuery(filterParams(filters))}`,
    ),
  operationalRisks: (filters: OperationsQueryParams = {}) =>
    apiFetch<OperationalRisks>(
      `${BASE}/operational-risks${toQuery(filterParams(filters))}`,
    ),
};

export const operationsQueryKeys = {
  all: ["operations"] as const,
  overview: (filters: OperationsQueryParams) =>
    [...operationsQueryKeys.all, "overview", filters] as const,
  inventory: (filters: OperationsQueryParams & { page?: number }) =>
    [...operationsQueryKeys.all, "inventory", filters] as const,
  suppliers: (filters: OperationsQueryParams) =>
    [...operationsQueryKeys.all, "suppliers", filters] as const,
  returns: (filters: OperationsQueryParams) =>
    [...operationsQueryKeys.all, "returns", filters] as const,
  warehouse: (filters: OperationsQueryParams) =>
    [...operationsQueryKeys.all, "warehouse", filters] as const,
  risks: (filters: OperationsQueryParams) =>
    [...operationsQueryKeys.all, "risks", filters] as const,
};

File: backend/app/schemas/operations.py
"""Operations Intelligence API response schemas."""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import PaginationMeta, ResponseMeta, utc_now


class OperationsMetric(BaseModel):
    id: str
    label: str
    value: float | int | None = None
    formatted_value: str | None = None
    unit: str | None = None
    delta: float | None = None
    delta_label: str | None = None
    trend: Literal["up", "down", "flat"] | None = None
    format: Literal["currency", "percent", "number"] | None = None
    available: bool = True
    source: str | None = None


class OperationsOverviewResponse(BaseModel):
    metrics: list[OperationsMetric]
    period_start: date | None = None
    period_end: date | None = None
    generated_at: datetime = Field(default_factory=utc_now)
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class InventoryRow(BaseModel):
    product: str
    product_id: str | None = None
    sku: str | None = None
    category: str | None = None
    stock: float | None = None
    safety_stock: float | None = None
    turnover: float | None = None
    inventory_value: float | None = None
    stock_status: str | None = None


class InventoryResponse(BaseModel):
    rows: list[InventoryRow]
    pagination: PaginationMeta
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class SupplierPerformanceRow(BaseModel):
    supplier: str
    supplier_id: str | None = None
    on_time_pct: float | None = None
    quality_score: float | None = None
    lead_time: float | None = None
    purchase_volume: float | None = None
    risk_level: str | None = None


class SupplierPerformanceResponse(BaseModel):
    rows: list[SupplierPerformanceRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class ReturnsRow(BaseModel):
    category: str
    return_count: float | None = None
    return_pct: float | None = None
    return_cost: float | None = None
    trend: Literal["up", "down", "flat"] | None = None


class ReturnsResponse(BaseModel):
    rows: list[ReturnsRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class WarehousePerformanceRow(BaseModel):
    warehouse: str
    inventory: float | None = None
    fulfillment: float | None = None
    stockouts: float | None = None
    average_processing_time: float | None = None


class WarehousePerformanceResponse(BaseModel):
    rows: list[WarehousePerformanceRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class OperationalRiskRow(BaseModel):
    risk: str
    severity: Literal["low", "medium", "high", "critical"] = "medium"
    owner: str | None = None
    recommendation: str | None = None


class OperationalRisksResponse(BaseModel):
    rows: list[OperationalRiskRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)
